File: swift/src/RemoteTransport.swift
import Foundation

public enum RemoteAuth: Sendable, Equatable {
    case none
    case basic(username: String, password: String)
    case bearerToken(String)
    case sshKey(username: String, privateKey: String, port: Int? = nil, strictHostKeyChecking: Bool = true)
    case sshAgent(username: String, port: Int? = nil, strictHostKeyChecking: Bool = true)
}

public struct TransportOptions: Sendable, Equatable {
    public var auth: RemoteAuth?
    public var insecureSkipTLSVerify: Bool

    public init(auth: RemoteAuth? = nil, insecureSkipTLSVerify: Bool = false) {
        self.auth = auth
        self.insecureSkipTLSVerify = insecureSkipTLSVerify
    }
}

struct ServiceAdvertisement {
    let refs: [RemoteRef]
    let capabilities: ServerCapabilities
}

private enum RemoteService {
    case uploadPack
    case receivePack

    var infoRefsService: String {
        switch self {
        case .uploadPack: return "git-upload-pack"
        case .receivePack: return "git-receive-pack"
        }
    }

    var requestContentType: String {
        switch self {
        case .uploadPack: return "application/x-git-upload-pack-request"
        case .receivePack: return "application/x-git-receive-pack-request"
        }
    }

    var resultContentType: String {
        switch self {
        case .uploadPack: return "application/x-git-upload-pack-result"
        case .receivePack: return "application/x-git-receive-pack-result"
        }
    }

    var commandName: String {
        switch self {
        case .uploadPack: return "git-upload-pack"
        case .receivePack: return "git-receive-pack"
        }
    }
}

private struct ParsedRemoteURL {
    let scheme: String
    let user: String?
    let host: String
    let port: Int?
    let path: String
}

func advertiseUploadPack(url: String, options: TransportOptions) throws -> ServiceAdvertisement {
    try advertiseRefs(url: url, service: .uploadPack, options: options)
}

func uploadPack(url: String, request: Data, options: TransportOptions) throws -> Data {
    try statelessRPC(url: url, service: .uploadPack, request: request, options: options)
}

func advertiseReceivePack(url: String, options: TransportOptions) throws -> ServiceAdvertisement {
    try advertiseRefs(url: url, service: .receivePack, options: options)
}

func receivePack(url: String, request: Data, options: TransportOptions) throws -> Data {
    try statelessRPC(url: url, service: .receivePack, request: request, options: options)
}

private func advertiseRefs(url: String, service: RemoteService, options: TransportOptions) throws -> ServiceAdvertisement {
    let parsed = try parseRemoteURL(url)
    let response: Data
    switch parsed.scheme {
    case "http", "https":
        response = try advertiseRefsHTTP(url: url, service: service, options: options)
    case "ssh":
        response = try advertiseRefsSSH(parsed: parsed, service: service, options: options)
    default:
        throw MuonGitError.invalidSpec("unsupported remote scheme '\(parsed.scheme)'")
    }

    let (decodedLines, _) = try unwrap(pktLineDecode(response))
    let lines = stripServicePreamble(decodedLines)
    let (refs, capabilities) = try unwrap(parseRefAdvertisement(lines))
    return ServiceAdvertisement(refs: refs, capabilities: capabilities)
}

private func stripServicePreamble(_ lines: [PktLine]) -> [PktLine] {
    guard lines.count >= 2 else { return lines }
    if case let .data(data) = lines[0],
       data.starts(with: Data("# service=".utf8)),
       case .flush = lines[1] {
        return Array(lines.dropFirst(2))
    }
    return lines
}

private func statelessRPC(url: String, service: RemoteService, request: Data, options: TransportOptions) throws -> Data {
    let parsed = try parseRemoteURL(url)
    switch parsed.scheme {
    case "http", "https":
        return try statelessRPCHTTP(url: url, service: service, request: request, options: options)
    case "ssh":
        return try statelessRPCSSH(parsed: parsed, service: service, request: request, options: options)
    default:
        throw MuonGitError.invalidSpec("unsupported remote scheme '\(parsed.scheme)'")
    }
}

private func advertiseRefsHTTP(url: String, service: RemoteService, options: TransportOptions) throws -> Data {
    var args = [
        "--silent",
        "--show-error",
        "--fail",
        "--location",
        "--header",
        "Accept: application/x-\(service.infoRefsService)-advertisement",
    ]
    applyHTTPAuthArgs(&args, options: options)
    args.append("\(trimTrailingSlash(url))/info/refs?service=\(service.infoRefsService)")
    return try runCommand(executable: "/usr/bin/curl", arguments: args, input: nil)
}

private func statelessRPCHTTP(url: String, service: RemoteService, request: Data, options: TransportOptions) throws -> Data {
    var args = [
        "--silent",
        "--show-error",
        "--fail",
        "--location",
        "--request",
        "POST",
        "--header",
        "Content-Type: \(service.requestContentType)",
        "--header",
        "Accept: \(service.resultContentType)",
        "--data-binary",
        "@-",
    ]
    applyHTTPAuthArgs(&args, options: options)
    args.append("\(trimTrailingSlash(url))\(serviceHTTPPath(service))")
    return try runCommand(executable: "/usr/bin/curl", arguments: args, input: request)
}

private func advertiseRefsSSH(parsed: ParsedRemoteURL, service: RemoteService, options: TransportOptions) throws -> Data {
    let remoteCommand = "\(service.commandName) --stateless-rpc --advertise-refs \(shellQuote(parsed.path))"
    return try runSSHCommand(parsed: parsed, options: options, remoteCommand: remoteCommand, input: nil)
}

private func statelessRPCSSH(parsed: ParsedRemoteURL, service: RemoteService, request: Data, options: TransportOptions) throws -> Data {
    let remoteCommand = "\(service.commandName) --stateless-rpc \(shellQuote(parsed.path))"
    return try runSSHCommand(parsed: parsed, options: options, remoteCommand: remoteCommand, input: request)
}

private func runSSHCommand(parsed: ParsedRemoteURL, options: TransportOptions, remoteCommand: String, input: Data?) throws -> Data {
    let sshConfig = try resolveSSHConfig(parsed: parsed, options: options)
    var args: [String] = ["-o", "BatchMode=yes"]
    if !sshConfig.strictHostKeyChecking {
        args.append(contentsOf: ["-o", "StrictHostKeyChecking=no", "-o", "UserKnownHostsFile=/dev/null"])
    }
    if let port = sshConfig.port {
        args.append(contentsOf: ["-p", String(port)])
    }
    if let privateKey = sshConfig.privateKey {
        args.append(contentsOf: ["-i", privateKey])
    }
    args.append(sshConfig.target)
    args.append(remoteCommand)
    return try runCommand(executable: "/usr/bin/ssh", arguments: args, input: input)
}

private func resolveSSHConfig(parsed: ParsedRemoteURL, options: TransportOptions) throws -> (target: String, port: Int?, privateKey: String?, strictHostKeyChecking: Bool) {
    switch options.auth {
    case nil, .some(.none):
        let target = parsed.user.map { "\($0)@\(parsed.host)" } ?? parsed.host
        return (target, parsed.port, nil, true)
    case let .some(.sshKey(username, privateKey, port, strictHostKeyChecking)):
        return ("\(username)@\(parsed.host)", port ?? parsed.port, privateKey, strictHostKeyChecking)
    case let .some(.sshAgent(username, port, strictHostKeyChecking)):
        return ("\(username)@\(parsed.host)", port ?? parsed.port, nil, strictHostKeyChecking)
    case .some(.basic), .some(.bearerToken):
        throw MuonGitError.auth("HTTP credentials cannot be used with ssh remotes")
    }
}

private func applyHTTPAuthArgs(_ args: inout [String], options: TransportOptions) {
    if options.insecureSkipTLSVerify {
        args.append("--insecure")
    }
    switch options.auth {
    case let .some(.basic(username, password)):
        args.append(contentsOf: ["--user", "\(username):\(password)"])
    case let .some(.bearerToken(token)):
        args.append(contentsOf: ["--header", "Authorization: Bearer \(token)"])
    default:
        break
    }
}

private func parseRemoteURL(_ url: String) throws -> ParsedRemoteURL {
    if !url.contains("://") {
        guard let colon = url.firstIndex(of: ":") else {
            throw MuonGitError.invalidSpec("invalid remote url '\(url)'")
        }
        let beforeColon = String(url[..<colon])
        let afterColon = String(url[url.index(after: colon)...])
        let (user, host) = parseUserHost(beforeColon)
        return ParsedRemoteURL(scheme: "ssh", user: user, host: host, port: nil, path: afterColon)
    }

    guard let schemeRange = url.range(of: "://") else {
        throw MuonGitError.invalidSpec("invalid remote url '\(url)'")
    }
    let scheme = String(url[..<schemeRange.lowerBound])
    let rest = String(url[schemeRange.upperBound...])
    let pathStart = rest.firstIndex(of: "/") ?? rest.endIndex
    let authority = String(rest[..<pathStart])
    let path = pathStart < rest.endIndex ? String(rest[pathStart...]) : "/"
    let userHost: (String?, String)
    if let at = authority.lastIndex(of: "@") {
        userHost = (String(authority[..<at]), String(authority[authority.index(after: at)...]))
    } else {
        userHost = (nil, authority)
    }
    let (host, port) = splitHostPort(userHost.1)
    return ParsedRemoteURL(scheme: scheme, user: userHost.0, host: host, port: port, path: path)
}

private func parseUserHost(_ text: String) -> (String?, String) {
    if let at = text.firstIndex(of: "@") {
        return (String(text[..<at]), String(text[text.index(after: at)...]))
    }
    return (nil, text)
}

private func splitHostPort(_ authority: String) -> (String, Int?) {
    if let colon = authority.lastIndex(of: ":") {
        let host = String(authority[..<colon])
        let portText = String(authority[authority.index(after: colon)...])
        if let port = Int(portText), portText.allSatisfy(\.isNumber) {
            return (host, port)
        }
    }
    return (authority, nil)
}

private func trimTrailingSlash(_ url: String) -> String {
    var value = url
    while value.hasSuffix("/") {
        value.removeLast()
    }
    return value
}

private func serviceHTTPPath(_ service: RemoteService) -> String {
    switch service {
    case .uploadPack: return "/git-upload-pack"
    case .receivePack: return "/git-receive-pack"
    }
}

private func shellQuote(_ value: String) -> String {
    "'\(value.replacingOccurrences(of: "'", with: "'\\''"))'"
}

private func unwrap<T>(_ result: Result<T, MuonGitError>) throws -> T {
    switch result {
    case let .success(value): return value
    case let .failure(error): throw error
    }
}

#if os(macOS)
private func runCommand(executable: String, arguments: [String], input: Data?) throws -> Data {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: executable)
    process.arguments = arguments

    let stdoutPipe = Pipe()
    let stderrPipe = Pipe()
    process.standardOutput = stdoutPipe
    process.standardError = stderrPipe

    let stdinPipe: Pipe?
    if input != nil {
        let pipe = Pipe()
        process.standardInput = pipe
        stdinPipe = pipe
    } else {
        stdinPipe = nil
    }

    try process.run()

    if let input, let stdinPipe {
        stdinPipe.fileHandleForWriting.write(input)
        stdinPipe.fileHandleForWriting.closeFile()
    }

    let output = stdoutPipe.fileHandleForReading.readDataToEndOfFile()
    let errorData = stderrPipe.fileHandleForReading.readDataToEndOfFile()
    process.waitUntilExit()

    guard process.terminationStatus == 0 else {
        let message = String(data: errorData, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "command failed"
        let lower = message.lowercased()
        if lower.contains("certificate") {
            throw MuonGitError.certificate(message)
        }
        if lower.contains("auth") || lower.contains("permission denied") || lower.contains("unauthorized") {
            throw MuonGitError.auth(message)
        }
        throw MuonGitError.invalid(message)
    }

    return output
}
#else
private func runCommand(executable: String, arguments: [String], input: Data?) throws -> Data {
    _ = executable
    _ = arguments
    _ = input
    throw MuonGitError.invalid("remote transport requires process support on this platform")
}
#endif

File: rust/src/remote_transport.rs
//! Remote transport client helpers for smart HTTP(S) and SSH.

use std::io::Write;
use std::process::{Command, Stdio};

use crate::error::MuonGitError;
use crate::transport::{
    parse_ref_advertisement, pkt_line_decode, PktLine, RemoteRef, ServerCapabilities,
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RemoteAuth {
    None,
    Basic {
        username: String,
        password: String,
    },
    BearerToken(String),
    SshKey {
        username: String,
        private_key: String,
        port: Option<u16>,
        strict_host_key_checking: bool,
    },
    SshAgent {
        username: String,
        port: Option<u16>,
        strict_host_key_checking: bool,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct TransportOptions {
    pub auth: Option<RemoteAuth>,
    pub insecure_skip_tls_verify: bool,
}

#[derive(Debug, Clone)]
pub struct ServiceAdvertisement {
    pub refs: Vec<RemoteRef>,
    pub capabilities: ServerCapabilities,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum RemoteService {
    UploadPack,
    ReceivePack,
}

impl RemoteService {
    fn http_info_refs_service(self) -> &'static str {
        match self {
            Self::UploadPack => "git-upload-pack",
            Self::ReceivePack => "git-receive-pack",
        }
    }

    fn http_content_type(self) -> &'static str {
        match self {
            Self::UploadPack => "application/x-git-upload-pack-request",
            Self::ReceivePack => "application/x-git-receive-pack-request",
        }
    }

    fn http_result_type(self) -> &'static str {
        match self {
            Self::UploadPack => "application/x-git-upload-pack-result",
            Self::ReceivePack => "application/x-git-receive-pack-result",
        }
    }

    fn command_name(self) -> &'static str {
        match self {
            Self::UploadPack => "git-upload-pack",
            Self::ReceivePack => "git-receive-pack",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct ParsedRemoteUrl {
    scheme: String,
    user: Option<String>,
    host: String,
    port: Option<u16>,
    path: String,
}

pub fn advertise_upload_pack(
    url: &str,
    opts: &TransportOptions,
) -> Result<ServiceAdvertisement, MuonGitError> {
    advertise_refs(url, RemoteService::UploadPack, opts)
}

pub fn upload_pack(
    url: &str,
    request: &[u8],
    opts: &TransportOptions,
) -> Result<Vec<u8>, MuonGitError> {
    stateless_rpc(url, RemoteService::UploadPack, request, opts)
}

pub fn advertise_receive_pack(
    url: &str,
    opts: &TransportOptions,
) -> Result<ServiceAdvertisement, MuonGitError> {
    advertise_refs(url, RemoteService::ReceivePack, opts)
}

pub fn receive_pack(
    url: &str,
    request: &[u8],
    opts: &TransportOptions,
) -> Result<Vec<u8>, MuonGitError> {
    stateless_rpc(url, RemoteService::ReceivePack, request, opts)
}

fn advertise_refs(
    url: &str,
    service: RemoteService,
    opts: &TransportOptions,
) -> Result<ServiceAdvertisement, MuonGitError> {
    let parsed = parse_remote_url(url)?;
    let response = match parsed.scheme.as_str() {
        "http" | "https" => advertise_refs_http(url, service, opts)?,
        "ssh" => advertise_refs_ssh(&parsed, service, opts)?,
        scheme => {
            return Err(MuonGitError::InvalidSpec(format!(
                "unsupported remote scheme '{}'",
                scheme
            )))
        }
    };

    let (lines, _) = pkt_line_decode(&response)?;
    let lines = strip_service_preamble(&lines);
    let (refs, capabilities) = parse_ref_advertisement(lines)?;
    Ok(ServiceAdvertisement { refs, capabilities })
}

fn stateless_rpc(
    url: &str,
    service: RemoteService,
    request: &[u8],
    opts: &TransportOptions,
) -> Result<Vec<u8>, MuonGitError> {
    let parsed = parse_remote_url(url)?;
    match parsed.scheme.as_str() {
        "http" | "https" => stateless_rpc_http(url, service, request, opts),
        "ssh" => stateless_rpc_ssh(&parsed, service, request, opts),
        scheme => Err(MuonGitError::InvalidSpec(format!(
            "unsupported remote scheme '{}'",
            scheme
        ))),
    }
}

fn advertise_refs_http(
    url: &str,
    service: RemoteService,
    opts: &TransportOptions,
) -> Result<Vec<u8>, MuonGitError> {
    let endpoint = format!(
        "{}{}?service={}",
        trim_trailing_slash(url),
        "/info/refs",
        service.http_info_refs_service()
    );
    let mut args = vec![
        "--silent".to_string(),
        "--show-error".to_string(),
        "--fail".to_string(),
        "--location".to_string(),
        "--header".to_string(),
        format!(
            "Accept: application/x-{}-advertisement",
            service.http_info_refs_service()
        ),
    ];
    apply_http_auth_args(&mut args, opts);
    args.push(endpoint);
    run_command("/usr/bin/curl", &args, None)
}

fn stateless_rpc_http(
    url: &str,
    service: RemoteService,
    request: &[u8],
    opts: &TransportOptions,
) -> Result<Vec<u8>, MuonGitError> {
    let endpoint = format!("{}{}", trim_trailing_slash(url), service_http_path(service));
    let mut args = vec![
        "--silent".to_string(),
        "--show-error".to_string(),
        "--fail".to_string(),
        "--location".to_string(),
        "--request".to_string(),
        "POST".to_string(),
        "--header".to_string(),
        format!("Content-Type: {}", service.http_content_type()),
        "--header".to_string(),
        format!("Accept: {}", service.http_result_type()),
        "--data-binary".to_string(),
        "@-".to_string(),
    ];
    apply_http_auth_args(&mut args, opts);
    args.push(endpoint);
    run_command("/usr/bin/curl", &args, Some(request))
}

fn advertise_refs_ssh(
    parsed: &ParsedRemoteUrl,
    service: RemoteService,
    opts: &TransportOptions,
) -> Result<Vec<u8>, MuonGitError> {
    let remote_cmd = format!(
        "{} --stateless-rpc --advertise-refs {}",
        service.command_name(),
        shell_quote(&parsed.path)
    );
    run_ssh_command(parsed, opts, &remote_cmd, None)
}

fn stateless_rpc_ssh(
    parsed: &ParsedRemoteUrl,
    service: RemoteService,
    request: &[u8],
    opts: &TransportOptions,
) -> Result<Vec<u8>, MuonGitError> {
    let remote_cmd = format!(
        "{} --stateless-rpc {}",
        service.command_name(),
        shell_quote(&parsed.path)
    );
    run_ssh_command(parsed, opts, &remote_cmd, Some(request))
}

fn run_ssh_command(
    parsed: &ParsedRemoteUrl,
    opts: &TransportOptions,
    remote_cmd: &str,
    input: Option<&[u8]>,
) -> Result<Vec<u8>, MuonGitError> {
    let mut args = Vec::new();

    let (username, port, strict_host_key_checking) = match opts.auth.as_ref() {
        Some(RemoteAuth::SshKey {
            username,
            private_key,
            port,
            strict_host_key_checking,
        }) => {
            args.push("-i".to_string());
            args.push(private_key.clone());
            (Some(username.clone()), port.or(parsed.port), *strict_host_key_checking)
        }
        Some(RemoteAuth::SshAgent {
            username,
            port,
            strict_host_key_checking,
        }) => (
            Some(username.clone()),
            port.or(parsed.port),
            *strict_host_key_checking,
        ),
        Some(RemoteAuth::None) | None => (parsed.user.clone(), parsed.port, true),
        Some(RemoteAuth::Basic { .. }) | Some(RemoteAuth::BearerToken(_)) => {
            return Err(MuonGitError::Auth(
                "HTTP credentials cannot be used with ssh remotes".into(),
            ))
        }
    };

    args.push("-o".to_string());
    args.push("BatchMode=yes".to_string());
    if !strict_host_key_checking {
        args.push("-o".to_string());
        args.push("StrictHostKeyChecking=no".to_string());
        args.push("-o".to_string());
        args.push("UserKnownHostsFile=/dev/null".to_string());
    }
    if let Some(port) = port {
        args.push("-p".to_string());
        args.push(port.to_string());
    }

    let target = match username {
        Some(user) => format!("{}@{}", user, parsed.host),
        None => parsed.host.clone(),
    };
    args.push(target);
    args.push(remote_cmd.to_string());

    run_command("/usr/bin/ssh", &args, input)
}

fn apply_http_auth_args(args: &mut Vec<String>, opts: &TransportOptions) {
    if opts.insecure_skip_tls_verify {
        args.push("--insecure".to_string());
    }

    match opts.auth.as_ref() {
        Some(RemoteAuth::Basic { username, password }) => {
            args.push("--user".to_string());
            args.push(format!("{}:{}", username, password));
        }
        Some(RemoteAuth::BearerToken(token)) => {
            args.push("--header".to_string());
            args.push(format!("Authorization: Bearer {}", token));
        }
        _ => {}
    }
}

fn run_command(
    command: &str,
    args: &[String],
    input: Option<&[u8]>,
) -> Result<Vec<u8>, MuonGitError> {
    let mut child = Command::new(command)
        .args(args)
        .stdin(if input.is_some() {
            Stdio::piped()
        } else {
            Stdio::null()
        })
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;

    if let Some(input) = input {
        let mut stdin = child.stdin.take().ok_or_else(|| {
            MuonGitError::Invalid("failed to open child stdin".into())
        })?;
        stdin.write_all(input)?;
    }

    let output = child.wait_with_output()?;
    if output.status.success() {
        Ok(output.stdout)
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
        let message = if stderr.is_empty() {
            format!("command '{}' failed", command)
        } else {
            stderr
        };

        if message.to_ascii_lowercase().contains("certificate") {
            Err(MuonGitError::Certificate(message))
        } else if message.to_ascii_lowercase().contains("auth")
            || message.to_ascii_lowercase().contains("permission denied")
            || message.to_ascii_lowercase().contains("unauthorized")
        {
            Err(MuonGitError::Auth(message))
        } else {
            Err(MuonGitError::Invalid(message))
        }
    }
}

fn parse_remote_url(url: &str) -> Result<ParsedRemoteUrl, MuonGitError> {
    if !url.contains("://") {
        let Some((user_host, path)) = url.split_once(':') else {
            return Err(MuonGitError::InvalidSpec(format!(
                "invalid remote url '{}'",
                url
            )));
        };
        let (user, host) = parse_user_host(user_host);
        return Ok(ParsedRemoteUrl {
            scheme: "ssh".into(),
            user,
            host: host.to_string(),
            port: None,
            path: path.to_string(),
        });
    }

    let (scheme, rest) = url
        .split_once("://")
        .ok_or_else(|| MuonGitError::InvalidSpec(format!("invalid remote url '{}'", url)))?;
    let path_start = rest.find('/').unwrap_or(rest.len());
    let authority = &rest[..path_start];
    let path = if path_start < rest.len() {
        &rest[path_start..]
    } else {
        "/"
    };

    let (user, host_port) = if let Some((user_info, host_part)) = authority.rsplit_once('@') {
        (Some(user_info.to_string()), host_part)
    } else {
        (None, authority)
    };

    let (host, port) = split_host_port(host_port);
    Ok(ParsedRemoteUrl {
        scheme: scheme.to_string(),
        user,
        host,
        port,
        path: path.to_string(),
    })
}

fn parse_user_host(user_host: &str) -> (Option<String>, &str) {
    if let Some((user, host)) = user_host.split_once('@') {
        (Some(user.to_string()), host)
    } else {
        (None, user_host)
    }
}

fn strip_service_preamble(lines: &[PktLine]) -> &[PktLine] {
    if matches!(lines.first(), Some(PktLine::Data(data)) if data.starts_with(b"# service="))
        && matches!(lines.get(1), Some(PktLine::Flush))
    {
        &lines[2..]
    } else {
        lines
    }
}

fn split_host_port(host_port: &str) -> (String, Option<u16>) {
    if let Some((host, port)) = host_port.rsplit_once(':') {
        if port.chars().all(|ch| ch.is_ascii_digit()) {
            return (host.to_string(), port.parse::<u16>().ok());
        }
    }
    (host_port.to_string(), None)
}

fn trim_trailing_slash(url: &str) -> &str {
    url.trim_end_matches('/')
}

fn service_http_path(service: RemoteService) -> &'static str {
    match service {
        RemoteService::UploadPack => "/git-upload-pack",
        RemoteService::ReceivePack => "/git-receive-pack",
    }
}

fn shell_quote(path: &str) -> String {
    format!("'{}'", path.replace('\'', "'\\''"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_remote_url_https() {
        let parsed = parse_remote_url("https://github.com/user/repo.git").unwrap();
        assert_eq!(parsed.scheme, "https");
        assert_eq!(parsed.host, "github.com");
        assert_eq!(parsed.port, None);
        assert_eq!(parsed.path, "/user/repo.git");
    }

    #[test]
    fn test_parse_remote_url_ssh_shorthand() {
        let parsed = parse_remote_url("git@github.com:user/repo.git").unwrap();
        assert_eq!(parsed.scheme, "ssh");
        assert_eq!(parsed.user.as_deref(), Some("git"));
        assert_eq!(parsed.host, "github.com");
        assert_eq!(parsed.path, "user/repo.git");
    }

    #[test]
    fn test_parse_remote_url_ssh_with_port() {
        let parsed = parse_remote_url("ssh://git@example.com:2222/repo.git").unwrap();
        assert_eq!(parsed.scheme, "ssh");
        assert_eq!(parsed.user.as_deref(), Some("git"));
        assert_eq!(parsed.host, "example.com");
        assert_eq!(parsed.port, Some(2222));
        assert_eq!(parsed.path, "/repo.git");
    }
}
